### PR TITLE
feat(cognito): user pool - account recovery

### LINKED ISSUE #8502: Enable 'AccountRecoverySetting' attribute on UserPool (L2 Construct)
### Description - New Feature Request
Enable 'AccountRecoverySetting' attribute on UserPool (L2 Construct)

### Reproduction Steps

```
const cfnUserPool = PHONE_ONLY_PASSWORD_LESS_USER_POOL.node.defaultChild as CfnUserPool;
    cfnUserPool.accountRecoverySetting = {
      recoveryMechanisms: [{
        name: 'admin_only',
        priority: 1
      }]
    }
```

### Environment

  - **CDK Framework Version:* 1.45.0 (build 0cfab15)*
  - **Node.js Version:*12.16* 
  - **OS               :*macOS*
  - **Language (Version):*TypeScript (~3.7.2)*

### PR DESCRIPTION
**Pull request by civilizeddev**
image

closes #8502

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cognito/README.md [hunk 1]
```diff
@@ -268,6 +268,17 @@ new UserPool(this, 'myuserpool', {
 
 Note that, `tempPasswordValidity` can be specified only in whole days. Specifying fractional days would throw an error.
 
+### Account Recovery Settings
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:271]: Let's make this a sub-section of security.

```suggestion
#### Account Recovery Settings
```

Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/README.md:271]: Add an equivalent entry to the TOC at the top of this file.

File: packages/@aws-cdk/aws-cognito/README.md [hunk 2]
```diff
@@ -268,6 +268,17 @@ new UserPool(this, 'myuserpool', {
...
+
+See [Recovering User Accounts](https://docs.aws.amazon.com/cognito/latest/developerguide/how-to-recover-a-user-account.html)
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/README.md:273]:
```suggestion
User pools can be configured on which method a user should use when recovering the password for their account. This
can either be email and/or SMS. Read more at [Recovering User
Accounts](https://docs.aws.amazon.com/cognito/latest/developerguide/how-to-recover-a-user-account.html)
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 3]
```diff
@@ -385,6 +385,47 @@ export interface EmailSettings {
...
+/**
+ * How will a user be able to recover their account?
+ *
+ * When a user forgets their password, they can have a code sent to their verified email or verified phone to recover their account.
+ * You can choose the preferred way to send codes below.
+ * We recommend not allowing phone to be used for both password resets and multi-factor authentication (MFA).
+ *
+ * @see https://docs.aws.amazon.com/cognito/latest/developerguide/how-to-recover-a-user-account.html
+ */
+export enum AccountRecoverySettings {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:397]: Just `AccountRecovery`

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 4]
```diff
@@ -509,6 +550,13 @@ export interface UserPoolProps {
    * @default true
    */
   readonly signInCaseSensitive?: boolean;
+
+  /**
+   * How will a user be able to recover their account?
+   *
+   * @default AccountRecoverySettings.PHONE_AND_EMAIL
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:557]: Let's not default to an option that is not recommended.

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 5]
```diff
@@ -908,6 +957,40 @@ export class UserPool extends UserPoolBase {
...
+        return {
+          recoveryMechanisms: [{ name: 'verified_phone_number', priority: 1 }],
+        };
+      case AccountRecoverySettings.NONE:
+        return {
+          recoveryMechanisms: [{ name: 'admin_only', priority: 1 }],
+        };
+      case AccountRecoverySettings.PHONE_AND_EMAIL:
+      default:
+        return undefined;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:991]: Throw exception if there's an option that is not implemented.

File: packages/@aws-cdk/aws-cognito/test/user-pool.test.ts [hunk 6]
```diff
@@ -1,17 +1,17 @@
 import '@aws-cdk/assert/jest';
 import { ABSENT } from '@aws-cdk/assert/lib/assertions/have-resource';
-import { Role } from '@aws-cdk/aws-iam';
+import * as iam from '@aws-cdk/aws-iam';
 import * as lambda from '@aws-cdk/aws-lambda';
-import { CfnParameter, Construct, Duration, Stack, Tag } from '@aws-cdk/core';
-import { Mfa, NumberAttribute, StringAttribute, UserPool, UserPoolIdentityProvider, UserPoolOperation, VerificationEmailStyle } from '../lib';
+import * as cdk from '@aws-cdk/core';
+import * as cognito from '../lib';
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/test/user-pool.test.ts:6]: Please revert these changes in this file. Use the same import style that already exists in the test.

Comment (civilizeddev) [reply packages/@aws-cdk/aws-cognito/test/user-pool.test.ts:6]: I see.

But it's still inconsistent.

```
import * as lambda from '@aws-cdk/aws-lambda';
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts (hunk 4 above)
Comment (civilizeddev) [reply packages/@aws-cdk/aws-cognito/lib/user-pool.ts:557]: Not recommended, but it's the default behavior on the console.

And it's also a default behavior on the cloudformation.

```
In the absence of this setting, Cognito uses the legacy behavior to determine the recovery method where SMS is preferred over email.
```

[AWS::Cognito::UserPool AccountRecoverySetting](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-userpool-accountrecoverysetting.html)

Comment (civilizeddev) [reply packages/@aws-cdk/aws-cognito/lib/user-pool.ts:557]: Alternatively, I could make it an essential property. But it would break the backward compatibility.

Comment (nija-at) [reply packages/@aws-cdk/aws-cognito/lib/user-pool.ts]: > Not recommended, but it's the default behavior on the console.
>
> And it's also a default behavior on the cloudformation.

The CDK can have its own default, separate from CloudFormation or the console.

File: packages/@aws-cdk/aws-cognito/test/user-pool.test.ts [hunk 7]
```diff
@@ -975,3 +975,54 @@ function fooFunction(scope: Construct, name: string): lambda.IFunction {
     handler: 'index.handler',
   });
 }
+
+test('AccountRecoverySetting should be configured correctly - when specified', () => {
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/test/user-pool.test.ts:979]: Move these tests to inside the `describe()` block.

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 8]
```diff
@@ -908,6 +957,42 @@ export class UserPool extends UserPoolBase {
...
+          recoveryMechanisms: [{ name: 'verified_phone_number', priority: 1 }],
+        };
+      case AccountRecovery.NONE:
+        return {
+          recoveryMechanisms: [{ name: 'admin_only', priority: 1 }],
+        };
+      case AccountRecovery.PHONE_AND_EMAIL:
+        return undefined;
+      default:
+        throw new Error('Invalid AccountRecovery type');
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:993]:
```suggestion
        throw new Error(`Unsupported AccountRecovery type - ${accountRecovery}`);
```

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 9]
```diff
@@ -509,6 +550,13 @@ export interface UserPoolProps {
    * @default true
    */
   readonly signInCaseSensitive?: boolean;
+
+  /**
+   * How will a user be able to recover their account?
+   *
+   * @default AccountRecovery.PHONE_AND_EMAIL
+   */
+  readonly accountRecovery?: AccountRecovery;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:559]: > Not recommended, but it's the default behavior on the console.
>
> And it's also a default behavior on the cloudformation.

The CDK can have its own default, separate from CloudFormation or the console.

File: packages/@aws-cdk/aws-cognito/lib/user-pool.ts [hunk 10]
```diff
@@ -908,6 +957,42 @@ export class UserPool extends UserPoolBase {
     }
     return schema;
   }
+
+  private accountRecovery(props: UserPoolProps): undefined | CfnUserPool.AccountRecoverySettingProperty {
+    const accountRecovery = (props.accountRecovery === undefined) ? AccountRecovery.PHONE_WITHOUT_MFA_AND_EMAIL : props.accountRecovery;
```
Comment (nija-at) [inline packages/@aws-cdk/aws-cognito/lib/user-pool.ts:962]:
```suggestion
    const accountRecovery = props.accountRecovery ?? AccountRecovery.PHONE_WITHOUT_MFA_AND_EMAIL;
```
